src/py/flwr/common/backoff.py

```diff
@@ -0,0 +1,278 @@
+        A constant value to yield or an iterable of such values.
+    """
+    if not isinstance(interval, Iterable):
+        interval = itertools.repeat(interval)
+    yield from interval
+
+
+def random_jitter(value: float) -> float:
+    """Jitter the value by a random number of milliseconds.
+
+    This adds up to 1 second of additional time to the original value.
+    Prior to backoff version 1.2, this was the default jitter behavior.
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return value + random.random()
+
+
+def full_jitter(value: float) -> float:
+    """Jitter the value across the full range (0 to value).
+
+    This corresponds to the "Full Jitter" algorithm specified in the
+    AWS blog post on the performance of various jitter algorithms.
+    See: http://www.awsarchitectureblog.com/2015/03/backoff.html
+
+    Parameters
+    ----------
+    value : float
+        The original, unjittered backoff value.
+    """
+    return random.uniform(0, value)
+
+
+@dataclass
+class Details:
+    """Details for event handlers in SafeInvoker."""
+
+    func: Callable[[Any], Any]
+    args: Tuple[Any, ...]
+    kwargs: Dict[str, Any]
+    tries: int
+    elapsed: float
+    exception: Optional[Exception] = None
+    wait: Optional[float] = None
+
+
+# pylint: disable-next=too-many-instance-attributes
+class SafeInvoker:
+    """Wrapper class for backoff and retry triggered by exceptions.
+
+    Parameters
+    ----------
+    wait: Generator[float, None, None]
+        A generator yielding successive wait times in seconds. If the generator
+        is finite, the giveup event will be triggered when the generator raises
+        `StopIteration`.
+    exception:  Union[Type[Exception], Iterable[Type[Exception]]]
+        An exception type (or iterable of types) that triggers backoff.
+    max_tries: Optional[int] (default: None)
+        The maximum number of attempts to make before giving up. Once exhausted,
+        the exception will be allowed to escape. If set to None, there is no limit
+        to the number of tries.
+    max_time: Optional[float] (default: None)
+        The maximum total amount of time to try before giving up. Once this time
+        has expired, the exception will be allowed to escape.
+    jitter: Optional[Callable[[float], float]] (default: full_jitter)
+        A function of the value yielded by `wait_gen` returning the actual time
+        to wait. This function helps distribute wait times stochastically to avoid
+        timing collisions across concurrent clients. Wait times are jittered by
+        default using the `full_jitter` function. To disable jittering, pass
+        `jitter=None`.
+    giveup_condition: Optional[Callable[[Exception], bool]] (default: None)
+        A function accepting an exception instance, returning whether or not
+        to give up. If set to None, the strategy is to always continue.
```

Comment:
In the last sentence, does `continue` mean to retry?

Reply:
It means the `SafeInvoker` will not give up prematurely but it can still give up due to all other reasons, i.e., `max_time`/`max_tries` is exceeded or an unexpected exception pops up.

I've updated the docstring for this argument.

Reply:
Looks good!